---
title: "How to use STTileMaker + an Example Analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{How to use STTileMaker + an Example Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(echo=TRUE, include = TRUE, warning=FALSE, message=FALSE,
  collapse = TRUE,
  comment = "#>"
)
```


## Goal of this Analysis

In my current rotation project within Dr. Jean Fan's lab, I am tasked with predicting gene expression levels by analyzing H&E-stained images through the lens of spatial transcriptomics (ST) data. To accomplish this, I have employed a technique known as [Hist2ST](https://pubmed.ncbi.nlm.nih.gov/35849101/), a method established in 2022 and widely recognized for its usefulness in Python environments. Despite initial challenges in deciphering and adapting the Hist2ST code for our data sets, I (finally) successfully ran Hist2ST on our data.

A pivotal component of my work involved developing a function to segment the ST data into distinct "tiles" according to the spatial positioning of each spot.

Recognizing that a significant portion of the ST research community prefers R for data analysis, I was motivated to enhance the accessibility and applicability of my workflow. This led to the creation of an analogous function in R, which I have aptly named STTileMaker.

The STTileMaker package is designed with a straightforward objective: to generate segmented tiles from H&E images that align with the precise locations of the spots within. Further, it can successfully read in all ST data to therefore be used in future downstream analysis. This capability sets the stage for a broad spectrum of subsequent analytical procedures, including the correlation of RGB values and gene expression, classification of spot characteristics, and the eventual integration of these tiles into an R-based algorithm poised to predict gene expression patterns. In the context of this data analysis, I will be demonstrating how gene expression correlates with RGB intensity within an H&E image.

Questions in the analysis then become:

1. What gene is most correlated to RGB intensities in an H&E image?

2. What gene is least correlated to RGB intensities in an H&E image? 

3. Does one channel predict better than others?

The data of interest is freely accessible from the [(Re)Building a Kidney Consortium database](https://www.atlas-d2k.org/chaise/record/#2/RNASeq:Experiment/RID=17-H1QC), offering a comprehensive spatial gene expression profile created using Space Ranger 1.2.1. The data encapsulates gene expression from the kidneys of an 8-10 week old C57BL6/J female mouse, specifically captured at the 12-hour mark following bilateral ischemia reperfusion injury. For detailed insights, including a data dictionary and the data's structure, I encourage visiting the provided link.

Due to the large size of the dataset, it was not possible to host it on GitHub. Nonetheless, I've included a snippet of code that, when executed, will facilitate the direct download of the dataset into a locally designated folder named "data." I hope this proves helpful!



## Reading in the Data

Now, lets begin the analysis. First we will read in all the necassary packages for the analysis.

```{r setup}
# read in functions
library(STTileMaker) # custom functions
library(ggplot2) # for plotting
library(tidyverse) # for data manipulation
library(here) # for file archiving
library(paletteer) # for colors
library(ggrepel) # for plotting
```


The code below is the code snippit that will gather the data from the website and save it to a folder called "data." This step is crucial for the analysis, as we are obtaining the dataset. However, you are free to use any ST dataset you would like for STTileMaker.

```{r}
# if file doesn't exist, download the data
# if (!file.exists(here("data/f12hr_140_processed/"))) {
if (!file.exists("~/Desktop/jhu/rotations/fan/STTileMaker/data/f12hr_140_processed/outs/")) {
  
  # message it wasn't found
  message("File not found, downloading now...")

  # create dir if needed
  # dir.create(here("data/"))
  dir.create("~/Desktop/jhu/rotations/fan/STTileMaker/data/")

  # URL of the file
  url = "https://www.atlas-d2k.org/hatrac/resources/rnaseq/study/17-E9P0/study_files/64ee25aacce94ff9a42850f158097813.tar.gz:JW6WAUMAZJ5TT3SF2LGG3C6V2U"

  # Destination file path (you can change this to your desired path and filename)
  # dest_file = here("data/downloaded_file.tar.gz")
  dest_file = "~/Desktop/jhu/rotations/fan/STTileMaker/data/downloaded_file.tar.gz"

  # Download the file
  download.file(url, dest_file, mode = "wb")

  # Assuming the file is named "downloaded_file.tar.gz" and is in the current working directory
  file_name = "downloaded_file.tar.gz"

  # Extract the file
  # untar(here("data/downloaded_file.tar.gz"), exdir = here("data/"))
  untar(dest_file, exdir = "~/Desktop/jhu/rotations/fan/STTileMaker/data/")

  # tarfile_list = untar(file_name, list = TRUE)
  # print(tarfile_list)

  # message when done
  message("Finished!")
}
```

Since we have all the data downloaded, we are ready to read it in with the "STTileMaker_read_data" function within the package. The object "data" within the code, which is our read in data, is a list of lists containing important features needed such as spot positions, cleaned expression data, etc. For all functions within the package please take a look at the docs section for more details.

After the data is read in, we can create the tiles from each spot's location by running the "STTileMaker_make_tiles" function. The output is a list of images, where each image is a "tile" or "patch" of the H&E image that contains a single spot. To ensure it worked successfully, we run the "STTileMaker_plot_tile" function to plot a single tile to view.

NOTE: Unfortunately, R-markdown would not allow me to knit this was the function here(). Therefore, if you would like to replicate this code you NEED to change the directory to where your data is located.

```{r, fig.align='center'}
# read in data
# data = STTileMaker_read_data(here::here("data/f12hr_140_processed/outs/"))
data =STTileMaker_read_data("~/Desktop/jhu/rotations/fan/STTileMaker/data/f12hr_140_processed/outs/")

# make tiles
tiles = STTileMaker_make_tiles(data)

# check to see it works
STTileMaker_plot_tile(tiles, 1)
STTileMaker_plot_tile(tiles, 100)
# it works!!
```

As the package continues to evolve, additional functions will be incorporated. However, the above selection represents the full range of functionalities available at this moment.

Now let's begin the analysis! 



## Data Analysis

```{r, fig.align='center', fig.width=7, fig.height=5}
# grab the mean of each red channel in each tile
features = map_dbl(tiles, ~ mean(.x[,,1])) # red channel

# set names of the features
names(features) = names(tiles)

# plot the histogram 
ggplot(data = data.frame(features), aes(x = features)) + 
  geom_histogram() +
  theme_linedraw() +
  labs(
    x = "Mean Value of Red Channel",
    y = "Count",
    title = "Mean Values of Tiles within the Red Channel",
    subtitle = str_wrap(paste0("Mean values of the red channel per tile. Note that it is fairly high for all tiles
                               located within tissue and is characterized by a normal distribution."), width = 100),
    caption = "Data Source: Kidney Consortium database"
  )  +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), plot.title = 
          element_text(face = "bold", hjust = 0.5), axis.text.x = element_text(size = 10))
```

The first plot shown above is a distribution of the mean values of the red channel. We see that it is normally distributed with a fairly high mean (range is 0-1). 

Next we see the distribution of correlation between the red channel and gene expression per tile. Unfortunately, we see that there is no strong correlation with any genes within the red channel. However, the red channel was found to be the best correlated channel with gene expression compared to blue and green. The other two channel results are not shown here, so I encourage you to try for yourself to confirm!

```{r, fig.align='center', fig.width=7, fig.height=5}
# normalize the data
normFactor = colSums(as.matrix(data$geneCountsClean))
mat = t(t(as.matrix(data$geneCountsClean))/normFactor)
mat = log10(mat * 1e6 +1)

# get correlation of each tile with gene expression
cors_list = map(rownames(mat), ~ cor(mat[.x, ], features))
# Convert the list to a numeric vector
cors = unlist(cors_list)


# plot the histogram 
ggplot(data = data.frame(cors), aes(x = cors)) + 
  geom_histogram() +
  labs(
    x = "Correlation",
    y = "Count",
    title = paste0("Mean Correlation between the Red Channel and Gene Expression"),
    subtitle = str_wrap(paste0("Correlation histogram of gene expression with the tissue and 
                               red channel values. Note the little deviation from 0, as only a few
                               genes are found to be correlated."), width = 100),
    caption = "Data Source: Kidney Consortium database"
  )  +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), plot.title = 
          element_text(face = "bold", hjust = 0.5), axis.text.x = element_text(size = 10))

```

Next, I wanted to see what the highest correlated gene and lowest correlated gene was when looking at the red channel. 

```{r, fig.align='center', fig.width=7, fig.height=5}
# get best correlated gene
# Assuming 'cors' is a named numeric vector
cors_df = tibble(name = rownames(mat), value = cors)

# Finding the names with the minimum 'cors' value
g_max = cors_df |>
  filter(value == max(value, na.rm = TRUE)) |>
  pull(name)

# get scaled expression of gene
ggexp = scale(mat[g_max,])[,1]
# get locations
df = data.frame(data$posClean, ggexp)


# Calculate the correlation coefficient
corr_coefficient = cor(mat[g_max, ], features)

ggplot() + 
  geom_point(aes(x = mat[g_max,], y = features)) +
  geom_smooth(aes(x = mat[g_max,], y = features), method = "lm") +
  theme_linedraw() +
  labs(
    x = "Log Gene Expression",
    y = "Mean value of Red Channel",
    title = paste0("Correlation of ",g_max," Expression and Mean Red Channel value"),
    subtitle = str_wrap(paste0("Mean expression per tile of gene ", g_max,". This is the most correlated gene
                               with the red channel values."), width = 100),
    caption = "Data Source: Kidney Consortium database"
  )  +
  annotate("text", x = -Inf, y = -Inf, label = sprintf("Correlation: %.2f", corr_coefficient), 
           hjust = -0.1, vjust = -.2, size = 5, color = "red") +  # customize theme
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), plot.title = 
          element_text(face = "bold", hjust = 0.5), axis.text.x = element_text(size = 10))

```

As we see, the highest correlated gene was Kap with a correlation of 0.41. While this is not particularly a strong correlation, it is better than nothing and serves as a great base for future analysis! We see there are lots of spots within the tissue that has 0 expression, with a big leap to a log expression of 2 or higher. Very interesting to see, and something that is worth diving into more to understand!

Now let's see what the highest correlated gene, Kap, looks like on the tissue.

```{r, fig.align='center', fig.width=7, fig.height=5}
# now we can plot with the image
x = grid::rasterGrob(data$img,
                      interpolate = FALSE,
                      width = grid::unit(1, "npc"),
                      height = grid::unit(1, "npc"))
# plot actual image
plt = ggplot() +
    annotation_custom(
        grob = x,
        xmin = 0,
        xmax = ncol(x$raster),
        ymin = 0,
        ymax = nrow(x$raster)) +
    coord_fixed(
        xlim = c(0, ncol(x$raster)),
        ylim = c(0, nrow(x$raster))) +
    theme_void()

# plot expression of gene
plt + geom_point(aes(x = x, y = y,
                     color = ggexp),
                 size = data$r/5,
                 data = df) +
    scale_colour_gradient2(
        low = "blue",
        mid = "white",
        high = "red") + 
  labs(
    title = paste0("Mean Expression of ",g_max," per Spot"),
                   color = "Mean Expression") +  # customize theme
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), plot.title = 
          element_text(face = "bold", hjust = 0.5))
```

This awesome figure allows us to see spatially where the gene is highly expressed and lowly expressed. Interestingly, this gene has high expression within the tissue with lower expression on the outside. We see essentially a gradient of expression of this gene, which is certainly a cool find. Remember that the red channel was able to do pretty well when comparing it to this genes expression with a correlation of 0.41.

Let's do the same thing with the least correlated gene!

```{r, fig.align='center', fig.width=7, fig.height=5}
# get worst correlated gene

# Finding the names with the minimum 'cors' value
g_min = cors_df |>
  filter(value == min(value, na.rm = TRUE)) |>
  pull(name)

# get scaled expression of gene
ggexp = scale(mat[g_min,])[,1]
# get locations
df = data.frame(data$posClean, ggexp)

# Calculate the correlation coefficient
corr_coefficient = cor(mat[g_min, ], features)

ggplot() + 
  geom_point(aes(x = mat[g_min,], y = features)) +
  geom_smooth(aes(x = mat[g_min,], y = features), method = "lm") +
  theme_linedraw() +
  labs(
    x = "Log Gene Expression",
    y = "Mean value of Red Channel",
    title = paste0("Correlation of ",g_min," Expression and Mean Red Channel value"),
    subtitle = str_wrap(paste0("Mean expression per tile of gene ", g_min,". This is the least correlated gene
                               with the red channel values."), width = 100),
    caption = "Data Source: Kidney Consortium database"
  )  +
  annotate("text", x = -Inf, y = -Inf, label = sprintf("Correlation: %.2f", corr_coefficient), 
           hjust = -0.1, vjust = -.2, size = 5, color = "red") +  # customize theme
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), plot.title = 
          element_text(face = "bold", hjust = 0.5), axis.text.x = element_text(size = 10))
```

With a correlation of -0.27, Igfbp5 was was the least correlated gene when comparing gene expression with red channel intensity. We see here a lot of tiles that did not contain any expression of this gene, with those tiles that did displaying a clear negative relationship.

Let's take a look at the gene expression on the tissue just to have an idea of what it would look like.

```{r, fig.align='center', fig.width=7, fig.height=5}
# plot expression of gene
plt + geom_point(aes(x = x, y = y,
                     color = ggexp),
                 size = data$r/5,
                 data = df) +
    scale_colour_gradient2(
        low = "blue",
        mid = "white",
        high = "red") + 
  labs(
    title = paste0("Mean Expression of ",g_min," per Spot"),
                   color = "Mean Expression") +  # customize theme
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), plot.title = 
          element_text(face = "bold", hjust = 0.5))
```

Well, it is of no surprise that it isn't very good! To me, it seems a bit sporadic and all over the place so we would likely not use this gene for any more downstream analysis.

Finally, I wanted to show one more feature I looked at in this analysis: the red channel mean expression minus the blue channel.

```{r, fig.align='center', fig.width=7, fig.height=5}
# grab the mean of each red channel in each tile
features = map_dbl(tiles, ~ mean(.x[,,1]) - mean(.x[,,3])) # red channel - blue channel

# set names of the features
names(features) = names(tiles)

# get correlation of each tile with gene expression
cors_list = map(rownames(mat), ~ cor(mat[.x, ], features))
# Convert the list to a numeric vector
cors = unlist(cors_list)

# get best correlated gene
# Assuming 'cors' is a named numeric vector
cors_df = tibble(name = rownames(mat), value = cors)

# Finding the names with the minimum 'cors' value
g_max = cors_df |>
  filter(value == max(value, na.rm = TRUE)) |>
  pull(name)

# get scaled expression of gene
ggexp_max = mat[g_max,]
# get locations
df = data.frame(data$posClean, ggexp_max)
# Use the rename function from dplyr
df = df |>
  rename(!!g_max := ggexp_max)

# Finding the names with the minimum 'cors' value
g_min = cors_df |>
  filter(value == min(value, na.rm = TRUE)) |>
  pull(name)

# get scaled expression of gene
ggexp_min = mat[g_min,]

# bind columns together
df = bind_cols(df,ggexp_min)
df = bind_cols(df,features)
# rename
names(df)[names(df) == "...4"] = g_min
names(df)[names(df) == "...5"] = "redminusblue"


# Transform data to long format
long_data = df |>
  gather(key = "gene", value = "expression", g_max, g_min)

# Create the plot with ggplot2
ggplot(long_data, aes(x = expression, y = redminusblue)) +
  geom_point() +
  geom_smooth(method = "lm") +
  facet_wrap(~ gene) +
  theme_linedraw() +
  labs(
    x = "Log Gene Expression",
    y = "Mean value of Red Channel minus Blue Channel",
    title = paste0("Mean Correlation between Red minus Blue Channel and Gene Expression"),
    subtitle = str_wrap(paste0("Two plots are displayed with the highest correlated gene, ", g_max,", with a 
                               correlation of ",round(max(cors_df$value, na.rm = TRUE), 2),
                               ", and the lowest correlated gene, ",g_min," with a correlation of
                               ",round(min(cors_df$value, na.rm = TRUE), 2),
                               "."), width = 95),
    caption = "Data Source: Kidney Consortium database"
  )  +
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), plot.title = 
          element_text(face = "bold", hjust = 0.5), axis.text.x = element_text(size = 10))
```

Oddly enough, we found that there were genes with a higher correlation when using the mean red channel values minus the blue channel values. The two shown above are the most highly correlated gene and the least correlated gene.


## Final Thoughts

In our data analysis demonstration, we illustrated the effective execution of all functions within STTileMaker, leveraging this tool to initiate a spatial transcriptomics (ST) data analysis, and presented some compelling findings. Moving forward, we aim to expand the package's capabilities by introducing additional functions and enhancements for more sophisticated analysis of ST tile data.



## Functions Used

*You must use at least six different functions from dplyr, tidyr, lubridate, stringr, or forcats:*

1. tibble() - tidyr
2. filter() - dplyr
3. pull() - dplyr
4. rename() - dplyr
5. bind_cols() - dplyr
6. gather() - tidyr
7. str_wrap() - stringr


*You must use at least two functions from purrr:*

1. map_dbl() - purrr
2. map() - purrr

*Your analysis should include at least three plots with you using at least three different geom_() functions from ggplot2 (or another package with geom_() functions)":*

1. geom_point() - ggplot2
2. geom_histogram() - ggplot2
3. geom_smooth() - ggplot2

*At least one plot should using a type of faceting (facet_grid() or facet_wrap()).:*

- Got it!








